File: 300.500/381.InsertDeleteGetRandomO1DuplicatesAllowed.cpp
#include <iostream>
#include <unordered_set>
#include <unordered_map>
#include <vector>
#include <stdlib.h>
#include <time.h>
using namespace std;

class RandomizedCollection {
public:
    /** Initialize your data structure here. */
    RandomizedCollection() {
        srand(time(nullptr));
    }
    
    /** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
    bool insert(int val) {
        data.insert(val);
        return data.count(val) == 1;
    }
    
    /** Removes a value from the collection. Returns true if the collection contained the specified element. */
    bool remove(int val) {
        auto it = data.find(val);
        if(it == data.end())
            return false;
        data.erase(it);
        return true;
    }
    
    /** Get a random element from the collection. */
    int getRandom() {
        int pos = rand()%data.size();
        auto it = data.begin();
        advance(it, pos);
        return *it;
    }
private:
    unordered_multiset<int> data;
};

class RandomizedCollection1 {
public:
    /** Initialize your data structure here. */
    RandomizedCollection1() {
        srand(time(nullptr));
    }
    
    /** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
    bool insert(int val) {
        if(size < max){
            data[size] = val;
            kvs[val].insert(size);
        }else{
            data.push_back(val);
            kvs[val].insert(max);
            ++max;
        }
        ++size;
        return kvs[val].size() == 1;
    }
    
    /** Removes a value from the collection. Returns true if the collection contained the specified element. */
    bool remove(int val) {
        auto it = kvs.find(val);
        if(it == kvs.end())
            return false;
        if(it->second.empty()){
            kvs.erase(it);
            return false;
        }
        auto pos = *it->second.begin();
        //交换data的数据
        swap(pos);
        if(it->second.empty())
            kvs.erase(it);
        --size;
        return true;
    }
    
    /** Get a random element from the collection. */
    int getRandom() {
        int pos = rand()%size;
        return data[pos];
    }
private:
    void swap(int pos){
        int end = size-1;
        if(pos == end || data[pos] == data[end]){
            kvs[data[pos]].erase(end);
            return;
        }
        kvs[data[pos]].erase(pos);
        data[pos] = data[end];
        kvs[data[end]].erase(end);
        kvs[data[end]].insert(pos);
    }
private:
    unordered_map<int, unordered_set<int>> kvs;
    vector<int> data;
    int max{};
    int size{};
};

int main(int argc, char const *argv[]){
    RandomizedCollection1 cls;
    cout << boolalpha << cls.insert(1) << " -> true" << endl;
    cout << boolalpha << cls.insert(1) << " -> false" << endl;
    cout << boolalpha << cls.insert(2) << " -> true" << endl;
    cout << boolalpha << cls.insert(1) << " -> false" << endl;
    cout << boolalpha << cls.insert(2) << " -> false" << endl;
    cout << boolalpha << cls.insert(2) << " -> false" << endl;
    cout << boolalpha << cls.remove(1) << " -> true" << endl;
    cout << boolalpha << cls.remove(2) << " -> true" << endl;
    cout << boolalpha << cls.remove(2) << " -> true" << endl;
    cout << boolalpha << cls.remove(2) << " -> true" << endl;
    cout << cls.getRandom() << endl;
    cout << cls.getRandom() << endl;
    return 0;
}

File: go/300.500/381.InsertDeleteGetRandomO1DuplicatesAllowed.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type RandomizedCollection struct {
	kvs  map[int]map[int]struct{}
	data []int
}

/** Initialize your data structure here. */
func Constructor() RandomizedCollection {
	rand.Seed(time.Now().UnixNano())
	return RandomizedCollection{
		kvs:  map[int]map[int]struct{}{},
		data: []int{},
	}
}

/** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
func (this *RandomizedCollection) Insert(val int) bool {
	l := len(this.data)
	this.data = append(this.data, val)
	if _, ok := this.kvs[val]; ok {
		this.kvs[val][l] = struct{}{}
	} else {
		this.kvs[val] = map[int]struct{}{}
		this.kvs[val][l] = struct{}{}
	}
	return len(this.kvs[val]) == 1
}

/** Removes a value from the collection. Returns true if the collection contained the specified element. */
func (this *RandomizedCollection) Remove(val int) bool {
	end := len(this.data) - 1
	if end < 0 {
		return false
	}
	if this.data[end] == val {
		delete(this.kvs[val], end)
		if len(this.kvs[val]) == 0 {
			delete(this.kvs, val)
		}
		this.data = this.data[:end]
		return true
	}
	if kv, ok := this.kvs[val]; ok {
		for pos := range kv {
			delete(kv, pos)
			this.data[pos] = this.data[end]
			delete(this.kvs[this.data[end]], end)
			this.kvs[this.data[end]][pos] = struct{}{}
			this.data = this.data[:end]
			break
		}
		if len(this.kvs[val]) == 0 {
			delete(this.kvs, val)
		}
		return true
	}
	return false
}

/** Get a random element from the collection. */
func (this *RandomizedCollection) GetRandom() int {
	pos := rand.Intn(len(this.data))
	return this.data[pos]
}

func main() {
	cls := Constructor()
	fmt.Println(cls.Insert(10), true)
	fmt.Println(cls.Insert(10), false)
	fmt.Println(cls.Insert(20), true)
	fmt.Println(cls.Insert(20), false)
	fmt.Println(cls.Insert(30), true)
	fmt.Println(cls.Insert(30), false)
	fmt.Println(cls.Remove(10), true)
	fmt.Println(cls.Remove(20), true)
	fmt.Println(cls.Remove(20), true)
	fmt.Println(cls.Remove(10), true)
	fmt.Println(cls.Remove(30), true)
	fmt.Println(cls.Insert(40), true)
	fmt.Println(cls.Remove(30), true)
	fmt.Println(cls.Remove(30), false)
	fmt.Println(cls.GetRandom())
	fmt.Println(cls.GetRandom())
}
